Add recipe for vscode-icons

taquangtrung opened the pull request:
### Brief summary of what the package does

This library displays beautiful SVG icons adopted from modern VSCode extensions.

### Direct link to the package repository

https://github.com/taquangtrung/vscode-icons.el

### Your association with the package

Maintainer.

### Relevant communications with the upstream package maintainer

N/A.

### Checklist

<!-- Please confirm by replacing `[]` with `[x]`: -->

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] I've used `M-x checkdoc` to check the package's documentation strings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

<!-- After submitting, please fix any problems the CI reports. -->

riscy commented:
Thanks for this.  Quick first pass -

⸺ `vscode-icons.el` with package-lint 20240923.1044:
```
3 issues found:
1960:5: error: You should depend on (emacs "29.1") if you need `js-ts-mode'.
1984:5: error: You should depend on (emacs "29.1") if you need `tsx-ts-mode'.
1987:5: error: You should depend on (emacs "29.1") if you need `typescript-ts-mode'.
```

Also, why does your recipe -
```elisp
(vscode-icons
    :fetcher github
    :repo "taquangtrung/vscode-icons.el"
    :files (:defaults "vscode-icons.png" "icons"))
```
include vscode-icons.png?  This doesn't seem to be referenced anywhere in code that I can see.

riscy commented:
> These issues are false alarms.

Agreed - you can disregard; I don't think there's anything you can do to suppress the lints on these.

I notice there have been quite a few code changes since last time; here's the latest, including some byte-compile nags -

⸺ `vscode-icons.el` with byte-compile using Emacs 29.4:
```
vscode-icons.el:1582:2: Warning: docstring wider than 80 characters
In vscode-icons-default-file-icon:
vscode-icons.el:1615:50: Warning: Unused lexical argument `size'
In vscode-icons-icon-for-dir:
vscode-icons.el:1619:2: Warning: docstring wider than 80 characters
In vscode-icons-default-dir-icon:
vscode-icons.el:1641:49: Warning: Unused lexical argument `size'
```

⸺ `vscode-icons.el` with checkdoc 0.6.2 (fix *within reason*):
```
vscode-icons.el:1570: Argument `icon-dir' should appear (as ICON-DIR) in the doc string
vscode-icons.el:1583: Some lines are over 80 columns wide
vscode-icons.el:1616: Argument `size' should appear (as SIZE) in the doc string
vscode-icons.el:1620: Some lines are over 80 columns wide
vscode-icons.el:1642: Argument `size' should appear (as SIZE) in the doc string
vscode-icons.el:1665: Argument `size' should appear (as SIZE) in the doc string
vscode-icons.el:1712: Argument `size' should appear (as SIZE) in the doc string
```

tbanel commented:
About those false `package-lint` warnings:

```
3 issues found:
1960:5: error: You should depend on (emacs "29.1") if you need `js-ts-mode'.
1984:5: error: You should depend on (emacs "29.1") if you need `tsx-ts-mode'.
1987:5: error: You should depend on (emacs "29.1") if you need `typescript-ts-mode'.
```
maybe you could use the hash syntax:

```lisp
(defvar vscode-icons-mode-icon-alist
  `((LaTeX-mode      ("vscode-icons" "file_tex.svg"))
    (c++-mode        ("vscode-icons" "file_cpp.svg"))
    …
    (,#'js-ts-mode   ("vscode-icons" "file_js.svg"))
    (js2-mode        ("vscode-icons" "file_js.svg"))
    …))
```
For me, this syntax silents `package-lint`.

Thierry (not a member of the Melpa team)
Have fun

mattiasb commented:
Not a legal expert by any means so I just want to check that
you've made sure that you fulfill the legal requirements for
including the icons that you do.

The reason that I'm asking is that [vscode-icons/vscode-icons][1]
uses [CC BY-SA][cc-by-sa] for its icons and I *believe* that you need to keep
them under that license (you release them under GPLv3 now it
seems). Also I think there's an attribution clause.

[microsoft/vscode-icons][2] had a slightly different CC license
FWIW and [material-extensions/vscode-material-icon-theme][3]
seems to use [MIT][license-mit].

[1]: https://github.com/vscode-icons/vscode-icons/tree/master?tab=readme-ov-file#license
[2]: https://github.com/microsoft/vscode-icons
[3]: https://github.com/material-extensions/vscode-material-icon-theme
[cc-by-sa]: https://creativecommons.org/licenses/by-sa/4.0/
[license-mit]: https://github.com/material-extensions/vscode-material-icon-theme/blob/main/LICENSE.md

riscy commented:
Thank you both!  @mattiasb this is a great point and something I think you should check on, @taquangtrung.

mattiasb commented:
Thanks, I just hope I wasn't discouraging @taquangtrung !

riscy commented:
Friendly ping, @taquangtrung :)

riscy commented:
I'll mark this awaiting-upstream, but feel free to reach out.

taquangtrung commented:
> Not a legal expert by any means so I just want to check that you've made sure that you fulfill the legal requirements for including the icons that you do.
> 
> The reason that I'm asking is that [vscode-icons/vscode-icons](https://github.com/vscode-icons/vscode-icons/tree/master?tab=readme-ov-file#license) uses [CC BY-SA](https://creativecommons.org/licenses/by-sa/4.0/) for its icons and I _believe_ that you need to keep them under that license (you release them under GPLv3 now it seems). Also I think there's an attribution clause.
> 
> [microsoft/vscode-icons](https://github.com/microsoft/vscode-icons) has a slightly different CC license FWIW and [material-extensions/vscode-material-icon-theme](https://github.com/material-extensions/vscode-material-icon-theme) seems to use [MIT](https://github.com/material-extensions/vscode-material-icon-theme/blob/main/LICENSE.md).
> 
> Mattias — Also not a Melpa dude :)

@mattiasb , @riscy: Thank you for the pointer!

I have checked the documentation of Melpa and it requires the package to be released under a GPL compatible license: https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org

However, both the CC BY-SA and MIT licenses seem not to be GPL-compatible. They are not listed here: https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses.

Currently, I don't know how to fix it.

taquangtrung commented:
@riscy: thanks for your comments!

I fixed all the linting and documentation issues.

I also restored the original GPL license, updated the README, and added integration to `Corfu` and `dired-mode`:  [vscode-icons-corfu.el](https://github.com/taquangtrung/vscode-icons.el/blob/main/vscode-icons-corfu.el) and  [vscode-icons-dired.el](https://github.com/taquangtrung/vscode-icons.el/blob/main/vscode-icons-dired.el).

For the suggestion of consolidating all SVG files into an archive file, this file would be updated every time a new icon is added. It would create extra overhead for Git traffic. Hence, I'd prefer to keep the list of SVG files for maintenance convenience.

Regarding the other package [vscode-icon-emacs](https://github.com/jojojames/vscode-icon-emacs), it only provides PNG icons. Adding SVG files into it is a major change to the package, hence, I decided to create a new one.

riscy commented:
> this file would be updated every time a new icon is added. It would create extra overhead for Git traffic. Hence, I'd prefer to keep the list of SVG files for maintenance convenience.

How often do you expecting new icons to be added (or changed)?  The only reason I'm mentioning this is because I have had traumatic ;) experiences with inefficient/outdated network storage and seen huge performance drops when working in directories containing tens of thousands of files (more than you have, but I still think 2000+ is getting there).

In the meantime, here's the latest -

⸺ `vscode-icons.el` with byte-compile using Emacs 30.1:
```
vscode-icons.el:1675:2: Warning: docstring wider than 80 characters
In end of data:
vscode-icons.el:1648:42: Warning: the function `cl-find-if' might not be defined at runtime.
```

⸺ `vscode-icons-dired.el` with package-lint 20250418.1425 and `package-lint-main-file` = "vscode-icons.el":
```
1 issue found:
8:0: error: Package-Requires outside the main file have no effect.
```

⸺ `vscode-icons-corfu.el` with package-lint 20250418.1425 and `package-lint-main-file` = "vscode-icons.el":
```
1 issue found:
8:0: error: Package-Requires outside the main file have no effect.
```

⸺ `vscode-icons.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- `;;; Commentary` is much wider than 80 characters
```

⸺ `vscode-icons.el` with checkdoc 0.6.2 (fix *within reason*):
```
vscode-icons.el:1635: First line is not a complete sentence
vscode-icons.el:1676: Some lines are over 80 columns wide
```

taquangtrung commented:
@riscy: Thank you for the pointer on the problem when handling a lot of files. Occasionally, I still need to add new icons for new file extensions, especially the less frequently used programming languages. Hopefully, the icon set will not grow exponentially. I'll try to keep them within 3000 icons.

Another reason I want to keep the icon files separately is that when new PRs come, I can easily review them.

I also fixed all the documentation and linting issues.

taquangtrung commented:
@riscy: Thank you for the pointer on the problem when handling a lot of files. Occasionally, I still need to add new icons for new file extensions, especially the less frequently used programming languages. Hopefully, the icon set will not grow exponentially. I'll try to keep them within 3000 icons.

Another reason I want to keep the icon files separately is  when new PRs come, I can easily review them.

I fixed all the documentation and linting issues.

riscy commented:
I understand your viewpoint.  You mentioned git traffic, but I'd also like to point out to you that (1) you're asking MELPA to distribute your package at scale - not some git SCM and (2) tar.gzipping your icons leads to a drop in size from 13M to 2.2M and (obviously) *thousands* fewer inodes taken up.  Re: PRs, you do have a point that working with uncompressed assets makes dev life easier.  But by indexing on a big package repo like MELPA, I feel you're signaling that you now care about *end users* more than developer users (and the amount of "stuff" you put on their machine).  Webpages minify js and imageoptim their visuals, finished games decompress stored assets on the fly, etc.  I really think this is the way to go.

I've said my piece and I won't insist, but I wanted to take one more stab at it while refreshing the checks -

First things first, I adjusted my "same package lookup" code and I now see https://melpa.org/#/vscode-icon - is there overlap here?  Is there a chance to join forces?

Second, package-lint shows some fresh issues -

⸺ `vscode-icons.el` with package-lint 20250418.1425 and `package-lint-main-file` = "vscode-icons.el":
```
2 issues found:
8:38: error: Package cl-seq is not installable.
38:10: warning: This file is not in the `cl-lib' ELPA compatibility package: require `cl-lib' instead.
```
This is happening because you have `;; Package-Requires: ((emacs "28.1") (cl-seq "2.02"))` - since you depend on Emacs 28, you don't need to `Package-Requires` cl-seq.  And - as the warning says - `(require cl-lib)` is the way to go.

taquangtrung commented:
@riscy: Somehow this PR was accidentally closed when I cleaned and force-pushed the master branch of my fork of Melpa to create other PRs. So, I re-opened it now.

The package https://melpa.org/#/vscode-icon that you mention provides a bunch of icon files in PNG format: https://github.com/jojojames/vscode-icon-emacs/tree/27cbf4f178924de1e2a09b4d87f87b5fa67c8cf4/icons/128. The quality of PNG icons will be affected when they are resized. Also, they do not support the transparent background well. This is the main reason that I created this new package to provide SVG icons.

My new package also supports new integration to other packages like Helm, Corfu, Treemacs, Dired, LSP: https://github.com/taquangtrung/vscode-icons.el

I'm considering your advice on packaging all SVG files into a tar file. Currently, I need to execute the external `tar` command, since the `tar-mode.el` package that you mentioned is mostly used to manipulate a tar buffer. Do you have any knowledge of an existing package that does the same so that I can learn from their design?

taquangtrung commented:
@riscy: Somehow this PR was accidentally closed when I cleaned and force-pushed the master branch of my fork of Melpa to create other PRs. So, I re-opened it now.

The package https://melpa.org/#/vscode-icon that you mention provides a bunch of icon files in PNG format: https://github.com/jojojames/vscode-icon-emacs/tree/27cbf4f178924de1e2a09b4d87f87b5fa67c8cf4/icons/128. The quality of PNG icons will be affected when they are resized. Also, they do not support the transparent background well. This is the main reason that I created this new package to provide SVG icons.

My new package also supports new integration to other packages like Helm, Corfu, Treemacs, Dired, LSP: https://github.com/taquangtrung/vscode-icons.el

I'm considering your advice on packaging all SVG files into a tar file. Currently, I need to execute the external `tar` command, since the `tar-mode.el` package that you mention mostly used to manipulate a tar buffer. Do you have any knowledge of an existing package that does the same so that I can learn from their design?

taquangtrung commented:
> > Somehow this PR was accidentally closed
> 
> I noticed and should've said something, just to check! Thanks for reopening.
> 
> > Do you have any knowledge of an existing package that does the same?
> 
> Not offhand although I could search. I think the function you want to use is the builtin tar-untar-buffer (for a given data buffer) - but if this is a hassle, let me know. (Maybe it suggests that we're missing a package for non-interactively working with archives, which seems surprising to me, but who knows!)

Yeah, I also think there is no native support for extracting a tar file, without having to load it to a buffer.

I also tried to implement a solution to extract icons using `tar-untar-buffer` like below (the commented code in: https://github.com/taquangtrung/vscode-icons.el/blob/9d55f5513338cd6a7b91f62bb572d6072bd86044/vscode-icons.el#L54-L57):

```emacs-lisp
  (let ((icons-tar-file (concat vscode-icons-root vscode-icons-icons-tar)))
    (with-current-buffer (find-file-noselect icons-tar-file)
      (tar-mode)
      (tar-untar-buffer)))
```

However, Emacs will throw exception at start-up if this code is called. Below is the screenshot when running `emacs --debug-init`:

![emcas-vscode-untar-error](https://github.com/user-attachments/assets/c3713e6f-ddfe-4284-b73d-8b04cf42b81e)

riscy commented:
This may not be the cleanest solution, so I'm definitely not going to insist.  You could also try something like this:
```elisp
(defun my-image-data (image-name)
  (save-window-excursion
    (let ((image-data nil)
          (tar-buffer (find-file "icons.tar.xz"))  ; changes current-buffer
          (img-buffer (tar--extract (tar-get-file-descriptor image-name))))
      (with-current-buffer img-buffer
        (setq image-data (buffer-substring (point-min) (point-max))))
      (kill-buffer tar-buffer)
      (kill-buffer img-buffer)
      image-data)))
```

And then once you have image data you can create/insert images as you please:
```elisp
(insert-image
 (create-image (my-image-data "icons/symbol-icons/dark/build.svg") 'svg t))
```

But let me know what you think.

taquangtrung commented:
> This may not be the cleanest solution, so I'm definitely not going to insist. You could also try something like this:
> 
> ```emacs-lisp
> (defun my-image-data (image-name)
>   (save-window-excursion
>     (let ((image-data nil)
>           (tar-buffer (find-file "icons.tar.xz"))  ; changes current-buffer
>           (img-buffer (tar--extract (tar-get-file-descriptor image-name))))
>       (with-current-buffer img-buffer
>         (setq image-data (buffer-substring (point-min) (point-max))))
>       (kill-buffer tar-buffer)
>       (kill-buffer img-buffer)
>       image-data)))
> ```
> 
> And then once you have image data you can create/insert images as you please:
> 
> ```emacs-lisp
> (insert-image
>  (create-image (my-image-data "icons/symbol-icons/dark/build.svg") 'svg t))
> ```
> 
> But let me know what you think.

Thanks for the suggestion on loading image data for each icon file. However, this might introduce overhead of opening the tar file each time an icon is loaded.

In the end, I found a solution to call `tar-untar-buffer` to extract all icons only once by implementing the function `vscode-icons-extract-icons` like below:

```emacs-lisp
(defun vscode-icons-extract-icons ()
  "Extract all icons from the tar file into the `icons' directory."
  (interactive)
  (message "vscode-icons: extracting icons files...")
  (let ((icons-tar-file (concat vscode-icons-root vscode-icons-icons-tar))
        (message-log-max nil)
        (inhibit-message t))
    (with-current-buffer (find-file-noselect icons-tar-file)
      (tar-untar-buffer)
      (kill-buffer))))
```

and invoke it under a timer of 0.1s to ensure other Emacs libraries are loaded before the tar file can be opened:

```emacs-lisp
(defun vscode-icons-enable-all ()
  "Enable all `vscode-icons' features."
  (interactive)

  ;; Extract icons using a timer to ensure the tar file is opened properly.
  (run-at-time 0.1 nil (lambda () (vscode-icons-extract-icons)))

  ;; Other code
  ;; ...
)

It works well for me now.

Can you help to review the remaining code using your checkers so that I can fix if there are any issues?

riscy commented:
Okay, sounds good!

I am running through the checks, and a couple things I want to mention:

First, are you aware of https://melpa.org/#/vscode-icon ?  If not a similar package, the name sure is (maybe we should consider disambiguating?)

Second, I think you have hard dependencies on things that you probably want to be soft dependencies - for example corfu?  This is causing some of the warnings below -

⸺ `vscode-icons.el` with byte-compile using Emacs 30.1:
```
vscode-icons.el:76:19: Warning: reference to free variable `corfu-margin-formatters'
vscode-icons.el:76:19: Warning: assignment to free variable `corfu-margin-formatters'
In end of data:
vscode-icons.el:92:6: Warning: the function `treemacs-load-theme' is not known to be defined.
vscode-icons.el:88:6: Warning: the function `vscode-icons-lsp-enable' is not known to be defined.
vscode-icons.el:84:6: Warning: the function `vscode-icons-helm-enable' is not known to be defined.
vscode-icons.el:76:45: Warning: the function `vscode-icons-corfu-formatter' is not known to be defined.
```

⸺ `vscode-icons-treemacs.el` with byte-compile using Emacs 30.1:
```
vscode-icons-treemacs.el:38:11: Error: Cannot open load file: No such file or directory, treemacs
```

⸺ `vscode-icons-lsp.el` with byte-compile using Emacs 30.1:
```
vscode-icons-lsp.el:32:11: Error: Cannot open load file: No such file or directory, lsp-icons
```

⸺ `vscode-icons-helm.el` with byte-compile using Emacs 30.1:
```
vscode-icons-helm.el:36:11: Error: Cannot open load file: No such file or directory, dash
```

⸺ `vscode-icons-core.el` with byte-compile using Emacs 30.1:
```
vscode-icons-core.el:1644:2: Warning: docstring wider than 80 characters
In vscode-icons-icon-for-mode:
vscode-icons-core.el:1700:36: Warning: Lexical argument shadows the dynamic variable mode-name
```

⸺ `vscode-icons.el` with package-lint 20250527.1845 and `package-lint-main-file` = "vscode-icons.el":
```
6 issues found:
8:38: error: Package cl-seq is not installable.
74:3: warning: `with-eval-after-load' is for use in configurations, and should rarely be used in packages.
78:3: warning: `with-eval-after-load' is for use in configurations, and should rarely be used in packages.
82:3: warning: `with-eval-after-load' is for use in configurations, and should rarely be used in packages.
86:3: warning: `with-eval-after-load' is for use in configurations, and should rarely be used in packages.
90:3: warning: `with-eval-after-load' is for use in configurations, and should rarely be used in packages.
```

⸺ `vscode-icons-treemacs.el` with package-lint 20250527.1845 and `package-lint-main-file` = "vscode-icons.el":
```
2 issues found:
41:0: error: "vit/create-icon" doesn't start with package's prefix "vscode-icons".
41:0: error: `vit/create-icon' contains a non-standard separator `/', use hyphens instead (see Elisp Coding Conventions).
```

⸺ `vscode-icons-lsp.el` with package-lint 20250527.1845 and `package-lint-main-file` = "vscode-icons.el":
```
1 issue found:
73:0: error: "vscode-icon-lsp-display-symbol-icon" doesn't start with package's prefix "vscode-icons".
```

⸺ `vscode-icons-helm.el` with package-lint 20250527.1845 and `package-lint-main-file` = "vscode-icons.el":
```
22 issues found:
134:0: error: "vih/advise-helm-make-source" doesn't start with package's prefix "vscode-icons".
134:0: error: `vih/advise-helm-make-source' contains a non-standard separator `/', use hyphens instead (see Elisp Coding Conventions).
186:1: warning: `with-eval-after-load' is for use in configurations, and should rarely be used in packages.
187:0: error: "vih/helm-ag--filter-one" doesn't start with package's prefix "vscode-icons".
187:0: error: `vih/helm-ag--filter-one' contains a non-standard separator `/', use hyphens instead (see Elisp Coding Conventions).
206:1: warning: `with-eval-after-load' is for use in configurations, and should rarely be used in packages.
207:0: error: "vih/helm-ff-prefix-filename" doesn't start with package's prefix "vscode-icons".
207:0: error: `vih/helm-ff-prefix-filename' contains a non-standard separator `/', use hyphens instead (see Elisp Coding Conventions).
225:1: warning: `with-eval-after-load' is for use in configurations, and should rarely be used in packages.
226:0: error: "vih/helm-imenu-icon-for-type" doesn't start with package's prefix "vscode-icons".
226:0: error: `vih/helm-imenu-icon-for-type' contains a non-standard separator `/', use hyphens instead (see Elisp Coding Conventions).
233:1: warning: `with-eval-after-load' is for use in configurations, and should rarely be used in packages.
234:0: error: "vih/helm-rg--make-face" doesn't start with package's prefix "vscode-icons".
234:0: error: `vih/helm-rg--make-face' contains a non-standard separator `/', use hyphens instead (see Elisp Coding Conventions).
249:1: warning: `with-eval-after-load' is for use in configurations, and should rarely be used in packages.
250:0: error: "vih/helm-xref-format-candidate" doesn't start with package's prefix "vscode-icons".
250:0: error: `vih/helm-xref-format-candidate' contains a non-standard separator `/', use hyphens instead (see Elisp Coding Conventions).
270:3: warning: `with-eval-after-load' is for use in configurations, and should rarely be used in packages.
272:3: warning: `with-eval-after-load' is for use in configurations, and should rarely be used in packages.
274:3: warning: `with-eval-after-load' is for use in configurations, and should rarely be used in packages.
276:3: warning: `with-eval-after-load' is for use in configurations, and should rarely be used in packages.
278:3: warning: `with-eval-after-load' is for use in configurations, and should rarely be used in packages.
```

⸺ `vscode-icons-core.el` with package-lint 20250527.1845 and `package-lint-main-file` = "vscode-icons.el":
```
2 issues found:
8:0: error: Package-Requires outside the main file have no effect.
38:10: warning: This file is not in the `cl-lib' ELPA compatibility package: require `cl-lib' instead.
```

⸺ `vscode-icons.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- vscode-icons.el#L80: It's safer to sharp-quote function names; use `#'`
```

taquangtrung commented:
@riscy: I changed the package name to `modern-icons.el` (https://github.com/taquangtrung/modern-icons.el/) and fixed all issues. Can you help to check again?

taquangtrung commented:
@riscy: FYI, I restructured the package and moved all integrations to existing Emacs packages into separate repositories to fix dependency issues. I also created a new organization to host them.

The core `modern-icons.el` library is now hosted at: https://github.com/emacs-modern-icons/modern-icons.el

Integrations to other Emacs libraries are hosted at separate repositories in: https://github.com/orgs/emacs-modern-icons/repositories

taquangtrung commented:
@riscy: oh, thanks for pointing it out. I fixed and committed to the `vscode-icons` branch of my Melpa fork, and didn't realize that I created the PR from the `master branch. I have merged it accordingly.

taquangtrung commented:
@riscy: I faced several problems when storing all the icons in a tar file:

- I can't track changes when editing/modifying the icons 
- When I'm editing icons and haven't packaged them into the tar file yet, if I open a new Emacs instance, it will automatically extract all the icons from the old tar file and override all my edits. 
- Since I can't keep track of the icons, sometimes I mistakenly package the draft icons or unnecessary files in the `icons` folder into the tar file.

These problems incur more effort to maintain and improve the icons. 

So, I decided to roll back to the previous method that keeps track of all the icons individually.

riscy commented:
Friendly ping. :)
